README.md
# Zam Test Library

**zam_test** is a typed version of the [test](https://pub.dev/packages/test) package.

## What's inside the package

Includes the following core components.

- [TestCase](https://pub.dev/documentation/zam_test/latest/zam_test/TestCase-class.html)
- [Test](https://pub.dev/documentation/zam_test/latest/zam_test/Test-class.html)
- [TestGroup](https://pub.dev/documentation/zam_test/latest/zam_test/TestGroup-class.html)

Check out all the components in detail [here](https://pub.dev/documentation/zam_test/latest/zam_test/zam_test-library.html)

## How to use

### Summary
- `Test` is the main class. It has two constructor methods, the `Test.single` and the `Test.multi`. If your test is simple and has only one case, then use `Test.single`.
- `TestCase` is something that is wrapped inside a `Test.multiple`. Avoid using a `TestCase` directly.
- `TestGroup` is used to run multiple `Test` together.

### Test

`Test` can be seen as the typed version of `test()` function in the original [test](https://pub.dev/packages/test) package. Mostly a `Test` will have multiple cases based on different inputs. For example, a login test has multiple test cases based on different inputs like wrong username, wrong password, correct username and password, etc. A `Test` can also have a single test case. Therefore, we have `MultiCasedTest` and `SingleCasedTest` respectively to address these. 

You can also create a new test by extending `Test` by following the steps given below.

- Create a class extending `Test`.
- Provide a `name`.
- Override `run` function which is called for every `TestCase`.
- Provide a list of `cases`.
- Override `intialize` and `dispose` when required.
- You can customize the `nameSuffix` and the `description` too.

```dart
class HeightTest extends Test<double, String> {
  @override
  final name = 'Height';

  @override
  run(input) {
    return Height(input).toStringInMetre();
  }

  @override
  final cases = [
    NegativeTestCase(
      when: 'Negative height value',
      input: -23,
      exception: HeightNotValidException,
    ),
    NegativeTestCase(
      when: 'Zero height value',
      input: 0,
      exception: HeightNotValidException,
    ),
    ValueTestCase(
      when: 'Positive Border height value',
      then: 'outputs value in m',
      input: 1,
      output: '0.01 m',
    ),
  ];
}

void main() {
  HeightTest().execute();
}
```

### TestGroup

`TestGroup` is used to run multiple `Test` together. It is more of a utility class. You can run tests without this.

```dart
void main() {
  TestGroup('BMI', [
    HeightTest(),
    // WeightTest(),
    // HeightTest(),
    // BmiCategoryTest(),
    // ...
    // ...
    // ... (you can add more test groups here)
  ]).execute();
}
```

### TestCase

It accepts two descriptive texts `when` and `then`, an `input`, a `matcher` and an optional callback called the `action`.

**CONSIDER**: Always consider executing it inside a `Test` class even when you want to run a single test case which is accomplished using `Test.single` constructor.

Simple way to execute a `TestCase` is as follows.

```dart
void main() {
  ValueTestCase(
    when: 'Positive Border height value',
    then: 'outputs value in m',
    input: 1.0,
    output: '0.01 m',
    action: (double input) => Height(input).toStringInMetre(),
  ).execute();
}
```

Currently we have provided the following test case types which derive from `TestCase`. We expect this list to grow in the future to handle various scenarios. Until then you can use `TestCase` since it accepts a custom matcher.

- [ValueTestCase](https://pub.dev/documentation/zam_test/latest/zam_test/ValueTestCase-class.html) - To match direct values.
- [NegativeTestCase](https://pub.dev/documentation/zam_test/latest/zam_test/NegativeTestCase-class.html) - To match exceptions.
- [BooleanTestCase](https://pub.dev/documentation/zam_test/latest/zam_test/BooleanTestCase-class.html) - To match booleans.

To learn more, move on to the [example section](https://pub.dev/packages/zam_test/example) or check out these dedicated [examples in github](https://github.com/zamstation/zam_test/blob/main/example/lib).

## Customization

You can override the following at the moment.

- `TestCase` -> `descriptionDelimiter` - Defaults to `' -> '`.
- `TestCase` -> description - It is generated by combining the `when` and `then` texts with a `descriptionDelimiter` in between.
- `Test` -> `nameSuffix` - Defaults to `':'`.
- `Test` -> `description` - It is generated by combining `name` and `nameSuffix`.
- `TestGroup` -> `nameSuffix` - Defaults to `' -'`.
- `TestGroup` -> `description` - It is generated by combining `name` and `nameSuffix`.

## Status
[![Build](https://github.com/zamstation/zam_test/actions/workflows/build_workflow.yaml/badge.svg)](https://github.com/zamstation/zam_test/actions/workflows/build_workflow.yaml)

## Contributors
- [Amsakanna](https://github.com/amsakanna)

## License
[BSD 3-Clause License](https://github.com/zamstation/zam_test/blob/main/LICENSE)

CHANGELOG.md
# 0.5.0
- Dependency Updates:
	- `flutter 3.0.0`
	- `dart 2.17.0`
	- `zam_core 0.7.0`
	- `lints 2.0.0`
	- `test 1.21.1`

# 0.4.0
- Dependency Updates: `zam_core 0.6.0`

# 0.3.0
- Modified `zam_test.dart` to export original test library.
- Dependency Updates:
	- `dart 2.13.0`
	- `zam_core 0.5.0`
	- `test 1.17.8`

# 0.2.0
- Added `BooleanTestCase`.

# 0.1.0
- Added new classes `Test`, `SingleCasedTest` and `MultiCasedTest`. Use `Test` as the main class instead of `TestCase` and `TestGroup`.
- Removed `TestRun`. Use `TestGroup` instead.
- Modified `TestCase` to make its action field optional.

# 0.0.2
- Dependency Updates: `zam_core 0.1.1`
- Fixes due to Dependency Updates: Fixed order of generics in callbacks.

# 0.0.1
- Includes the core classes `TestCase`, `TestGroup` and `TestRun`.
- Includes `ValueTestCase` for straight value matches.
- Includes `NegativeTestCase` for exception matches.
- Includes `ParameterizedTypeMatcher`.
- Includes utilities `sleep` and `shutDown`.


example/lib/main.dart
import 'package:zam_test/utilities.dart';
import 'package:zam_test/zam_test.dart';

///
/// [TestGroup] is used to run multiple [Test] together.
///
void main() {
  TestGroup('BMI', [
    HeightTest(),
    HeightCanCreateTest(),
    // ...
    // ...
    // ... (you can add more tests here)
  ]).execute();

  // Alternatively, you can write as follows.
  // HeightTest().execute();
  // But [TestGroup] will help you when you have many [Test]s.
}

///
/// Create a class extending [Test].
/// Provide a [name]. (If not, defaults to class name. In this case, it is 'HeightTest')
/// You can optionally override the [nameSuffix] and the [description].
/// Override [initialize] and [dispose] when required.
/// Override [run] which is a function called for every [TestCase].
/// Provide a list of [testCases].
///
class HeightTest extends Test<double, String> {
  @override
  final name = 'Height';

  /// Runs once before the first [TestCase] is run.
  /// [sleep] is an utility function in this package.
  @override
  initialize() async {
    await sleep(seconds: 1);
  }

  /// This function is run for every [TestCase].
  @override
  run(input) {
    return Height(input).toStringInMetre();
  }

  /// Runs once after the last [TestCase] is run.
  /// [shutdown] is an utility function in this package.
  @override
  dispose() async {
    await shutdown(seconds: 1);
  }

  /// Provide all the [TestCase]s here in an array.
  ///
  /// Here are the different types of test cases.
  ///   [TestCase] - Base class which expects an input and a matcher.
  ///   [ValueTestCase] - Expects an input and an output.
  ///   [NegativeTestCase] - Expects an input and an exception type.
  ///   [BooleanTestCase] - Expects an input and output is set based on constructor.
  @override
  final cases = [
    NegativeTestCase(
      when: 'Negative height value',
      input: -23,
      exception: HeightNotValidException,
    ),
    NegativeTestCase(
      when: 'Negative Border height value',
      input: -1,
      exception: HeightNotValidException,
    ),
    NegativeTestCase(
      when: 'Zero height value',
      input: 0,
      exception: HeightNotValidException,
    ),
    ValueTestCase(
      when: 'Positive Border height value',
      then: 'outputs value in m',
      input: 1,
      output: '0.01 m',
    ),
    ValueTestCase(
      when: 'Positive height value',
      then: 'outputs value in m',
      input: 45,
      output: '0.45 m',
    ),
    ValueTestCase(
      when: 'Positive Extreme height value',
      then: 'outputs value in m',
      input: 834,
      output: '8.34 m',
    ),
  ];
}

class HeightCanCreateTest extends Test<double, bool> {
  @override
  bool run(double input) {
    return Height.canCreate(input);
  }

  @override
  final cases = [
    BooleanTestCase.truthy(
      when: 'Positive height value',
      input: 45,
    ),
    BooleanTestCase.falsy(
      when: 'Negative height value',
      input: -22,
    ),
  ];
}

class Height {
  final double value;
  double get valueInMetre => value / 100;

  static bool canCreate(double value) => value > 0;

  Height(double value)
      : value = value > 0 ? value : throw HeightNotValidException(value);

  @override
  String toString() {
    return '${value.round()} cm';
  }

  String toStringInMetre() {
    return '${valueInMetre.toStringAsFixed(2)} m';
  }
}

class HeightNotValidException implements Exception {
  final double value;

  const HeightNotValidException(this.value);

  @override
  String toString() {
    return '$runtimeType: $value is an invalid height value. Height should be positive.';
  }
}
